internal/server/server.go

```diff
@@ -1,24 +1,168 @@
+		cookie, err := r.Cookie(s.cookieName)
+		if err != nil {
+			logAndError(w, http.StatusUnauthorized, err, "error reading cookie")
+			return
+		}
+
+		// Get claims from token
+		claims, err := jwt.TokenClaims(cookie.Value)
+		if err != nil {
+			logAndError(w, http.StatusUnauthorized, err, "error reading claims from jwt token")
+			return
+		}
+
+		if claims.Subject == "" {
+			logAndError(w, http.StatusUnauthorized, err, "sub claim is empty")
+			return
+		}
+
+		if len(claims.Groups) == 0 {
+			logAndError(w, http.StatusUnauthorized, err, "groups claim is empty")
+			return
+		}
+
+		// Make Subject claim the value used for login
+		login := claims.Subject
+
+		log.Infof("user %s is attempting to log in", login)
+		log.Debugf("claim groups for user %s: %v", login, claims.Groups)
+
+		// validUserGroups represents the intersection of user groups from claim with the group
+		// mapping in configuration
+		validUserGroups := config.ValidUserGroups(claims.Groups, s.groups)
+		log.Debugf("valid user groups for user %s: %v", login, validUserGroups)
+		if len(validUserGroups) == 0 {
+			logAndError(w, http.StatusUnauthorized, err, "no user groups matching configured mapping")
+			return
+		}
+
+		// lookup the user globally first as if it is not present it would need to
+		// be created
+		orgUser, err := s.grafanaClient.LookupUser(claims.Subject)
+		if err != nil {
+			logAndError(w, http.StatusUnauthorized, err, "error looking for user")
+			return
+		}
+
+		// Assign the email from claim so it's always updated
+		orgUser.Email = claims.Email
+
+		// if the Login field in user is empty, it means that the user wasn't found
+		if orgUser.Login == "" {
+			log.Infof("no user with login %s found, creating new one", login)
+			orgUser.Login = login
+
+			uid, err := s.grafanaClient.CreateUser(orgUser)
+			if err != nil {
+				logAndError(w, http.StatusUnauthorized, err, "error creating new user")
+				return
+			}
+
+			orgUser.ID = uid
+		}
+
+		// Mapping of role per org
+		userOrgsRole := make(map[int64]models.RoleType)
+
+		for _, groupConfig := range validUserGroups {
+			if groupConfig.GrafanaAdmin != orgUser.IsAdmin {
+				log.Infof("updating global admin status for user %s", login)
+				err := s.grafanaClient.UpdateUserPermissions(orgUser.ID, groupConfig.GrafanaAdmin)
+				if err != nil {
+					logAndError(w, http.StatusUnauthorized, err, "error assigning the user as Grafana admin")
+					return
+				}
+			}
+
+			for _, org := range groupConfig.Orgs {
+				// Check if the users has a more permissive role and apply that instead
+				if !grafana.IsRoleAssignable(userOrgsRole[org.ID], models.RoleType(org.Role)) {
+					continue
+				}
+
+				userOrgsRole[org.ID] = models.RoleType(org.Role)
+			}
+		}
+
+		for orgID, role := range userOrgsRole {
+			err = s.grafanaClient.UpsertOrgUser(orgID, orgUser, string(role))
+			if err != nil {
+				logAndError(w, http.StatusUnauthorized, err, "error upserting user")
+				return
+			}
```

Comment:
I understand what you're saying. Here are the 2 options I see:

- Have a default policy of allowing the user even if cannot be upserted and use the default values in Grafana for auto org number and auto org assign role.
- Handle default org and role logic inside the proxy, so we can compare with explicit values and let it through if they match, like default orgid 1 and default role Viewer.
- Treat the current behaviour as a known error, so users won't be authorized if the don't have a valid org assigned in our config. Basically without hidden defaults but all explicit in config. If the org is incorrect then we would need to fix it, same reason as why there's no automatic org management handling in the proxy.

Reply:
I like a combination of those options.  Support a default behavior in the auth proxy config map.

Reply:
We do a GetUser -> CreateUser flow to create the 'global' grafana user on lines 91-112 so if we make it here we have a global grafana user alreadt. I assume this user is by default viewer in the default org but that should be confirmed. This produces the minimum needed for minting the session token and grafana authentication so it should be safe to log these errors and continue.

If a group -> role mapping fails here it will result in that user either not having access to an org or having less permission than they should. Either way, they should still be able to authenticate to grafana with missing permissions vs getting an 'unauthorized' error.

Reply:
for point 1, the assumption of default viewer in default org is incorrect, I mean, as I mentioned on my 1st and 2nd points that can either happen after Grafana has "realized" the new user, that meaning that the proxy won't know about those default values and will simply allow the user to passthrough and let Grafana handle defaults. Or the 2nd option is to move that logic to the proxy so we say what org and role are defaults and explicitly assign it, if there's no other Org that matches or as default value for all.

Reply:
![Screen Shot 2021-11-23 at 3 17 33 PM](https://user-images.githubusercontent.com/53450727/143103326-e14d6b6a-6232-4499-8bdf-93bde11d4550.png)

If there are 0 successful groups upserts and we still let the user in it would have the above permission.